src/emu/emuopts.h

```diff
@@ -499,14 +502,12 @@ class emu_options : public core_options
 	virtual void command_argument_processed() override;
 
 private:
-	struct software_options
-	{
-		std::unordered_map<std::string, std::string>    slot;
-		std::unordered_map<std::string, std::string>    image;
-	};
+	struct software_options;
 
 	// slot/image/softlist calculus
-	software_options evaluate_initial_softlist_options(const std::string &software_identifier);
+	software_options evaluate_initial_softlist_options(std::string_view software_identifier);
+	std::optional<software_options> evaluate_single_software(const machine_config &config, device_type_enumerator<software_list_device, software_list_device> &swdevice_iter, software_list_device &swlistdev, std::string_view software_name) const;
```

Comment:
Since the function always iterates over all software list devices, is there really a good reason to pass in the enumerator here?  You could just pass in the root of the part of the device tree that should be searched and construct the enumerator locally – they’re cheap to create.  Passing by reference like this also suggests that the function may pass state back to the caller by modifying it.